### ISSUE TITLE
Server example?

### ISSUE THREAD

**Issue by Azeirah** [ggml-org/whisper.cpp#1369]
I'm working on a voice-controlled application and I want to run small .wav files through whisper fairly often.

What I noticed is that it takes almost 50% of the total time just to load the model every single time I run `./main -m ... "my-short-spoken-command.wav"`

I think it'd be nice if like in llama.cpp this project includes a server example so the model only has to be loaded once and stays in memory after loading.

**Comment by FSSRepo** [comment]
Maybe, when I finish working on optimizing stable-diffusion.cpp and adding a server to it, I could create a server example for whisper.cpp.

**Comment by Azeirah** [comment]
I posted the code as a PR https://github.com/ggerganov/whisper.cpp/pull/1375

**Comment by ggerganov** [comment]
Hey all, I notice several server examples being proposed. This is super cool!

I'm planning to do a major update to `whisper.cpp` in the following few days, bringing some new features and performance improvements. This will be the highest priority, so for the sake of less distractions, the server examples will have to wait until we finish the new release. Sorry for the delay

**Comment by ggerganov** [comment]
Hi again! I think we should restart the server efforts now that `v1.5.0` is released.

I like both #1375 and #1380, so not sure how to decide which one to integrate. 
@Azeirah and @felrock (and others): do you have any opinion on this.

Also, I think we should aim to support the OpenAI Audio API for speech to text: https://platform.openai.com/docs/api-reference/audio

The approach in #1418 is also interesting, so it can be merged as an alternative solution to the REST-based server example.

**Comment by felrock** [comment]
Hello! I'm keen on fixing and merging my changes for the server. I've seen that the use for server in llama.cpp has enabled projects such as ollama and others. So i think it's an important application to have, for users to easily create interfaces against.

I have also started to create a similar server solution for [bark.cpp](https://github.com/PABannier/bark.cpp) because in my use case I would like to have some sort of voice(a bit more granular than espeak). Which would complete the full llm robot, a brain(llama.cpp), ears(whisper.cpp) and voice(bark.cpp).

**Comment by ggerganov** [comment]
Yup, I agree that a server can find many interesting applications.

> Which would complete the full llm robot, a brain(llama.cpp), ears(whisper.cpp) and voice(bark.cpp).

Yes! Great idea - we are getting close :)

**Comment by colinator** [comment]
Also agree. To hawk my proposal [#1418](https://github.com/ggerganov/whisper.cpp/pull/1418) (that fork is a bit messy, but something like it) - I think it'd be really great to have the ability to create many types of servers. For instance, I might want a gRPC server. Or a rest server. Or a ROS pub-sub node. Likewise: many types of encodings for the result: maybe json, maybe bson, maybe protobuf, etc. I think it'd require very little refactoring - basically, just the core stream server as a class with a method `infer(audio data*)`. Happy to help!

**Comment by ggerganov** [comment]
First pass of a server example has been merged (#1380).

Looks like streaming and diarization are 2 of the most requested features for the server. Not sure if we can do something meaningful for diarization, but we should able to provide a streaming API relatively easy.

**Comment by felrock** [comment]
I left the diarization parameters in there so it might be working, I didn't know how it worked or how to test it.

**Comment by nortekax** [comment]
The server works well, but when speech is short, like "lights on", "lights off", etc, _it doesn't produce any text_.

I suspect `whisper.cpp` needs a long context because the `command` example asks for a long sentence first, before it can work properly. The `command`  example tells the user:
```
process_general_transcription: Say the following phrase: 'Ok Whisper, start listening for commands.'
```

I think a way to provide a _context_ for the server (like `command` does) would be useful to provide agents that need short commands, like "lights on", "lights off", etc.

**Comment by felrock** [comment]
Ok, I've tried sending single word .wav files to the server and have it respond with the correct work. Did you try using the prompt flag? Should be something similar to what you describe.

**Comment by ggerganov** [comment]
It should work with short audio too. The prompt can help in some situations to make the transcript more robust, but is not required in general.

**Comment by Azeirah** [comment]
> The server works well, but when speech is short, like "lights on", "lights off", etc, _it doesn't produce any text_.
> 
> I suspect `whisper.cpp` needs a long context because the `command` example asks for a long sentence first, before it can work properly. The `command` example tells the user:
> 
> ```
> process_general_transcription: Say the following phrase: 'Ok Whisper, start listening for commands.'
> ```
> 
> I think a way to provide a _context_ for the server (like `command` does) would be useful to provide agents that need short commands, like "lights on", "lights off", etc.

I haven't tested this specific server implementation, but the server implementation I was using previously definitely did work with short commands, I specifically made it for that purpose.

So either

- There's something wrong with your data
- There's something wrong with the glue between your data and the server
- There's something wrong with the server implementation

Have you tried it with longer audio?

**Comment by nortekax** [comment]
> Have you tried it with longer audio?

Longer audio always works; the same problem that happens with `server` also happens with `main`. Only `command` works well with the short voice commands, but only after you say the first long sentence that `command` asks you to say.

**Comment by nortekax** [comment]
From @felrock 
> Ok, I've tried sending single word .wav files to the server and have it respond with the correct work. Did you try using the prompt flag? Should be something similar to what you describe.

From @ggerganov 
> It should work with short audio too. The prompt can help in some situations to make the transcript more robust, but is not required in general.

I am using the following (based on https://stackoverflow.com/questions/30006609/using-sox-for-voice-detection-and-streaming) to generate the wav: 

```
sox -q -c 1 -r 16000 -d  -b 16 -r 16000 "$outwav"  silence 1 0.3 1% 1 0.3 1%
```
I play the wav and it is okay. How do you generate your wavs ?

**Comment by felrock** [comment]
Im using a python script, which uses PyAudio. I record for about three seconds per .wav file

**Comment by nortekax** [comment]
thanks, how do you detect voice?

**Comment by nortekax** [comment]
Okay, I solved the problem. Post it here for those interested. I simply _padded the wav file with 500ms of silence at the beginning and 1 second of silence at the end,_ and everything works fine now.

**Comment by nortekax** [comment]
It would be really useful to add  the grammar/commands.txt  functionality in the `command` example.

**Comment by ggerganov** [comment]
Ah sorry, about that - I forgot there is logic to ignore sub one second audio:

https://github.com/ggerganov/whisper.cpp/blob/a5881d619c8440c0e0f226be15fc5ab0eec5f9bb/whisper.cpp#L5193-L5198

**Comment by bobqianic** [comment]
> Ah sorry, about that - I forgot there is logic to ignore sub one second audio:
> 
> https://github.com/ggerganov/whisper.cpp/blob/a5881d619c8440c0e0f226be15fc5ab0eec5f9bb/whisper.cpp#L5193-L5198

BTW, why we need this logic here?
